### ISSUE TITLE
Tracked multilabel metrics are wrong

### ISSUE THREAD

**Issue by ATriantafyllopoulos** [autrainer/autrainer#57]
The bug is here https://github.com/autrainer/autrainer/blob/2cdd8bb64ca6550a69a58e22bf291664c6d2a9f0/autrainer/training/training.py#L814

and is caused by the use of `average='micro'` or `average='macro'` with `sklearn` when computed for each label separately. I have not yet figured out exactly why this happens (i.e. what goes on internally) but results are very inflated compared to the "normal" computation of the metric when using all target columns (these are the results that are printed).

This has potentially impacted the results reported in https://huggingface.co/autrainer/edansa-2019-cnn10-32k-t

**Comment by ramppdev** [comment]
Oh very bad, how did you notice?

Is this bug related to https://github.com/autrainer/autrainer/pull/53, not right? It was present before?
If so, then the model https://huggingface.co/autrainer/edansa-2019-cnn10-32k-t also has to be retrained.

**Comment by ATriantafyllopoulos** [comment]
I will investigate and come up with a fix. No need for retraining, it's only the reporting of the final metrics that are affected, not that of the intermediate ones upon which the model state is made

**Comment by ramppdev** [comment]
Yes the metrics for selecting the best state (and tracked in the `metrics.csv`) are correct:
https://github.com/millinma/-ModularAITaskTraining/blob/d3a4b5c4e80dd5afbdb30a22c8cb30b7c9f46623/autrainer/training/training.py#L797-L798

The problem is that we cannot calculate the scores per class this way right?
For the model we still need to adjust the reported performance values in the model card.

**Comment by ATriantafyllopoulos** [comment]
Exactly, I think we can solve this by bypassing the value for 'average' or using some reshaping. And yes, for the public model we just have to adapt the reported performance..

**Comment by ramppdev** [comment]
I think the problem is that the metrics do not really make sense when evaluated in a per class manner.
We should also in general probably not just take the average but actually compute the `all` entry for disaggregated evaluation.

**Comment by ATriantafyllopoulos** [comment]
Can you please add the predictions in the [cnn10-edansa](https://huggingface.co/autrainer/edansa-2019-cnn10-32k-t/tree/main/_test) model?

**Comment by ramppdev** [comment]
> Can you please add the predictions in the [cnn10-edansa](https://huggingface.co/autrainer/edansa-2019-cnn10-32k-t/tree/main/_test) model?

The predictions on the test set?

**Comment by ATriantafyllopoulos** [comment]
> The predictions on the test set? You mean update `dev.yaml`, `test_holistic.yaml` and the metadata in the model card once #58 is merged?

Well, I think it's safer to add the predictions as well, so that we can correct the `yamls` but also have the data there for future reference (and corrections if needed)

**Comment by ramppdev** [comment]
Sounds good, i will retrain the model once #58 is merged and update the metadata.

**Comment by AlexanderGbd** [comment]
> and is caused by the use of average='micro' or average='macro' with sklearn when computed for each label separately.

There is also the option to set `average=None` as they say:

```
average{‘micro’, ‘macro’, ‘samples’, ‘weighted’, ‘binary’} or None, default=’binary’

    This parameter is required for multiclass/multilabel targets. If None, the scores for each class are returned.
```

Isn't this also what we want for the multilabel case?

**Comment by ATriantafyllopoulos** [comment]
So the difference can be seen in the following snippet:

```python
>>> x = [0, 0, 0, 1, 1, 1]
>>> y = [0, 1, 1, 0, 1, 1]
>>> sklearn.metrics.f1_score(x, y, average="binary")
0.5714285714285714
>>> sklearn.metrics.f1_score(x, y, average=None)
array([0.4       , 0.57142857])
```

Basically, `average='binary'` computes the `f1_score` only for the positive label (default: 1). `average=None` computes it for both 0 and 1 and would lead to a different result (if we average).

In the case of multilabel-classification, we might have a lot of datapoints for which one particular label is 0 and the prediction is also 0 (this is because we have many non-overlapping labels). This means that the f1-score for the negative label (0) is very high, and this boosts performance by a lot. I think `average='binary'` is the safest option for this case as it will only compute the performance when the label is active, so precision/recall will only be computed for those datapoints.

What do you think?

**Comment by AlexanderGbd** [comment]
Okay I guess I understand what you mean. So we want to focus only on positive labels in order to see how well the model detects active labels, ignoring the inactive labels. As there is indeed label sparsity with many 0s in multi-label tasks this makes definitely sense. 

For a complete "label-level analysis" we can then still use `average=None` if we also wanted to get a feeling on that.

**Comment by AlexanderGbd** [comment]
I just tested both `average=binary` and `average=None` when evaluating some predictions and it produced the same results. 
See the following example with 6 classes:

```
x = np.array(
    [[0, 0, 1, 0, 0, 1], 
     [0, 1, 1, 0, 1, 0]]
)
y = np.array(
    [[0, 1, 1, 0, 1, 1], 
    [0, 1, 0, 1, 0, 0]]
)

class_wise_f1 = f1_score(y, x, average=None)
print("Class-wise: ", class_wise_f1.tolist())

f1_scores_binary = [f1_score(y[:, i], x[:, i], average="binary") for i in range(y.shape[1])]
print("Binary: ", f1_scores_binary)

>>> Class-wise:  [1.0, 0.6666666666666666, 0.6666666666666666, 0.0, 0.0, 1.0]
>>> Binary:  [1.0, 0.6666666666666666, 0.6666666666666666, 0.0, 0.0, 1.0]
```

**Comment by ATriantafyllopoulos** [comment]
> I just tested both `average=binary` and `average=None` when evaluating some predictions and it produced the same results. See the following example with 6 classes:
> 
> ```
> x = np.array(
>     [[0, 0, 1, 0, 0, 1], 
>      [0, 1, 1, 0, 1, 0]]
> )
> y = np.array(
>     [[0, 1, 1, 0, 1, 1], 
>     [0, 1, 0, 1, 0, 0]]
> )
> 
> class_wise_f1 = f1_score(y, x, average=None)
> print("Class-wise: ", class_wise_f1.tolist())
> 
> f1_scores_binary = [f1_score(y[:, i], x[:, i], average="binary") for i in range(y.shape[1])]
> print("Binary: ", f1_scores_binary)
> 
> >>> Class-wise:  [1.0, 0.6666666666666666, 0.6666666666666666, 0.0, 0.0, 1.0]
> >>> Binary:  [1.0, 0.6666666666666666, 0.6666666666666666, 0.0, 0.0, 1.0]
> ```

Yes, the changes are only relevant when the predictions & targets are 1d, can you please check the differences again?

**Comment by ramppdev** [comment]
Is this solved by now?

**Comment by ramppdev** [comment]
> Sounds good, i will retrain the model once #58 is merged and update the metadata.

@ATriantafyllopoulos i have updated the metrics and added the results CSVs for the predictions.

**Comment by ATriantafyllopoulos** [comment]
Closing as stale
